packages/flutter_tools/lib/src/device.dart

```diff
@@ -210,103 +275,133 @@ abstract class DeviceManager {
+    if ((mustBeSupportedByFlutter && !device.isSupported()) ||
+        (mustBeSupportedForProject && !isDeviceSupportedForProject(device)) ||
+        (mustBeSupportedForAll && !await isDeviceSupportedForAll(device))) {
+      return false;
     }
+    return true;
+  }
 
-    return devices;
+  Future<bool> isDeviceSupportedForAll(Device device) async {
+    // User has specified `--device all`.
+    //
+    // Always remove web and fuchsia devices from `--all`. This setting
+    // currently requires devices to share a frontend_server and resident
+    // runner instance. Both web and fuchsia require differently configured
+    // compilers, and web requires an entirely different resident runner.
+    final TargetPlatform devicePlatform = await device.targetPlatform;
+    if (device.isSupported() &&
+        devicePlatform != TargetPlatform.fuchsia_arm64 &&
+        devicePlatform != TargetPlatform.fuchsia_x64 &&
+        devicePlatform != TargetPlatform.web_javascript &&
+        isDeviceSupportedForProject(device)) {
+      return true;
+    }
+    return false;
   }
 
   /// Returns whether the device is supported for the project.
   ///
-  /// This exists to allow the check to be overridden for google3 clients. If
+  /// A device can be supported by Flutter but not supported for the project
+  /// (e.g. when the user has removed the iOS directory from their project).
+  ///
+  /// This also exists to allow the check to be overridden for google3 clients. If
   /// [flutterProject] is null then return true.
-  bool isDeviceSupportedForProject(Device device, FlutterProject? flutterProject) {
+  bool isDeviceSupportedForProject(Device device) {
+    if (!device.isSupported()) {
+      return false;
+    }
     if (flutterProject == null) {
       return true;
     }
-    return device.isSupportedForProject(flutterProject);
+    return device.isSupportedForProject(flutterProject!);
+  }
+}
+
+/// A class for filtering devices.
+///
+/// If [mustBeConnected] is true, only devices detected as connected will be included.
+///
+/// If [supportFilter] is provided, only devices matching the requirements will be included.
+///
+/// If [deviceConnectionFilter] is provided, only devices matching the DeviceConnectionInterface will be included.
+class DeviceDiscoveryFilter {
+  DeviceDiscoveryFilter({
+    this.mustBeConnected = true,
+    this.supportFilter,
+    this.deviceConnectionFilter,
+  });
+
+  final bool mustBeConnected;
+  final DeviceDiscoverySupportFilter? supportFilter;
+  final DeviceConnectionInterface? deviceConnectionFilter;
+
+  Future<bool> matchesRequirements(Device device) async {
+    final DeviceDiscoverySupportFilter? localSupportFilter = supportFilter;
+    if ((mustBeConnected && !device.isConnected) ||
+        (localSupportFilter != null &&
+            !await localSupportFilter.matchesRequirements(device)) ||
+        !matchesDeviceConnectionInterface(device, deviceConnectionFilter)) {
```

Comment:
nit. I think pulling some of these sub-expressions into named variables would make this easier to read:

```suggestion
    final bool deviceIsNotConnectedButMustBe = mustBeConnected && !device.isConnected;
    final bool deviceFailsFilter = localSupportFilter != null && !(await localSupportFilter.matchesRequirements(device));
    final bool doesNotMatchConnectionInterface = !matchesDeviceConnectionInterface(device, deviceConnectionFilter);
    if (deviceIsNotConnectedButMustBe || deviceFailsFilter || doesNotMatchConnectionInterface) {
```